onecodex/models/collection.py

```diff
@@ -389,6 +391,171 @@ def taxonomy(self):
+            "reaction",
+        ]
+        if annotation not in valid_annotation_values:
+            raise ValueError(f"'annotation' must be one of {', '.join(valid_annotation_values)}")
+        if annotation == "pathways":
+            if metric not in ["coverage", "abundance"]:
+                raise ValueError(
+                    "if using annotation='pathways', 'value' must be one of ['coverage', 'abundance']"
+                )
+        elif metric not in ["cpm", "rpk"]:
+            raise ValueError(
+                f"if using annotation={annotation}, 'value' must be one of ['cpm', 'rpk']"
+            )
+
+        data = {}
+        all_features = set()
+        # iterate over functional profiles, subset data, and store in data dict
+        for profile in self._functional_profiles:
+            # get table from mainline
+            table = profile.table(annotation=annotation, taxa_stratified=taxa_stratified)
+            # filter by indicated metric
+            if not taxa_stratified:
+                table = table[table["metric"] == "total_" + metric]
+            else:
+                table = table[table["metric"] == metric]
+            # if taxa stratified, concatenate id and taxon_name
+            if taxa_stratified:
+                table["id"] = table["id"] + "_" + table["taxon_name"]
+            # store tables for later retrieval
+            data[profile.id] = dict(zip(table.id, table.value))
+            all_features.update(set(table["id"]))
+
+        features_to_ix = {}
+        feature_list = []
+        for ix, feature in enumerate(all_features):
+            features_to_ix[feature] = ix
+            feature_list.append(feature)
+
+        # initialize an array and fill it
+        array = np.full(shape=(len(data), len(features_to_ix)), fill_value=np.nan)
+        profile_ids = []
+        for profile_index, profile_id in enumerate(data):
+            for feature_id in data[profile_id]:
+                array[profile_index, features_to_ix[feature_id]] = data[profile_id][feature_id]
+            profile_ids.append(profile_id)
+        df = pd.DataFrame(
+            array, index=pd.Index(profile_ids, name="functional_profile_uuid"), columns=feature_list
+        )
+
+        if fill_missing:
+            df.fillna(filler, inplace=True)
+        self._cached["functional_results"] = df
+        self._cached["functional_results_content"] = {
+            "annotation": annotation,
+            "taxa_stratified": taxa_stratified,
+            "metric": metric,
+            "fill_missing": fill_missing,
+            "filler": filler,
+        }
+
+    def _functional_results(self, **kwargs):
+        if "functional_results" not in self._cached:
+            self._collate_functional_results(**kwargs)
+        # check for diff in kwargs from previous call
+        for k in kwargs:
+            if self._cached["functional_results_content"].get(k) != kwargs[k]:
+                self._collate_functional_results(**kwargs)
```

Comment:
this will call `_collate_functional_results()` once for each kwarg that differs, when it really only needs to be called once. also, if `**kwargs` is a *subset* of the cached kwargs, it may not trigger a cache refresh when it should have

Reply:
Oh, good catch, this is sloppy.

Fixed, added tests.
@jairideout I'm curious to hear your feedback on the way the defaults for the args are defined both in the caller and the function definition. This feels wrong to me. Maybe I should take the defaults out of the function definition and just "patch" the kwargs passed by the caller `_functional_results()`? Would appreciate your advice on a better pattern here.

Reply:
see https://github.com/onecodex/onecodex/pull/424#discussion_r919136680 for an idea I had